File: svelte/src/lib/server/mock/state.ts
import type {
  ActivityLog,
  AppContext,
  DashboardData,
  NotificationBadges,
} from "$lib/types";
import { faker } from "@faker-js/faker";
import { nanoid } from "nanoid";
import { createDeckSmall } from "./decks";
import { createLessonSmall } from "./lessons";
import { createTaskSmall } from "./tasks";
import { createProfile, createStudent, createUser } from "./user";

export function createAppContext(
  overrides: Partial<AppContext> = {},
): AppContext {
  return {
    user: createUser(),
    profile: createProfile(),
    callUrl: null,
    students: faker.helpers.multiple(createStudent, { count: 5 }),
    preferences: {
      autoSubscribe: false,
      emailNotifications: false,
      pushNotifications: false,
      theme: "light",
      language: "ru",
      ...overrides,
    },
  };
}

export function createDashboardData(
  overrides: Partial<DashboardData> = {},
): DashboardData {
  return {
    tasks: faker.helpers.multiple(createTaskSmall),
    lessons: faker.helpers.multiple(createLessonSmall),
    decks: faker.helpers.multiple(createDeckSmall),
    activity: faker.helpers.multiple(createActivity),
    learnData: {
      cardsStudiedToday: faker.number.int({ min: 3, max: 100 }),
      currentStreak: faker.number.int({ min: 1, max: 30 }),
    },
    ...overrides,
  };
}

function createActivity(): ActivityLog {
  return {
    action: faker.helpers.arrayElement(["updated", "new", "deleted"]),
    createdAt: faker.date.recent().toISOString(),
    modelId: nanoid(),
    modelType: faker.helpers.arrayElement(["lesson", "deck", "task"]),
  };
}

export function createBadges(): NotificationBadges {
  return {
    unseenDecks: faker.number.int({ min: 0, max: 15 }),
    unseenLessons: faker.number.int({ min: 0, max: 15 }),
    unseenTasks: faker.number.int({ min: 0, max: 15 }),
    dueCards: faker.number.int({ min: 0, max: 15 }),
  };
}

File: axum/src/services/types/state.rs
use crate::types::{
    ActivityLog, DeckSmall, LessonSmall, Profile, SimpleStats, Student, TaskSmall, User,
    UserPreferences,
};
use serde::Serialize;
use utoipa::ToSchema;

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct AppContext {
    pub user: User,
    pub profile: Profile,
    pub call_url: Option<String>,
    pub students: Vec<Student>,
    pub preferences: UserPreferences,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct DashboardData {
    pub tasks: Vec<TaskSmall>,
    pub lessons: Vec<LessonSmall>,
    pub decks: Vec<DeckSmall>,
    pub activity: Vec<ActivityLog>,
    pub learn_data: SimpleStats,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct NotificationBadges {
    pub unseen_tasks: i64,
    pub unseen_lessons: i64,
    pub unseen_decks: i64,
    pub due_cards: Option<i64>,
}
